Clarify `wait(::Task)` documentation

Seelengrab opened the pull request:
Small follow-up to https://github.com/JuliaLang/julia/pull/53685, since having critical documentation for a specific method be only mentioned in a catch-all docstring isn't quite as helpful as it could be (e.g. when looking at inline documentation with LSP, where it happens to know that the variable being passed is a `Task`). In addition, this also documents that `wait` on a `Task` will throw if given `current_task`.

I've also split the function into a prototype and the actual implementation, so that the same "more specific docs at the method definition" transform can be applied for `GenericCondition` too.

Seelengrab committed:
Clarify `wait(::Task)` documentation

Seelengrab commented:
Failure is real, it's due to the change in exception type - https://buildkite.com/julialang/julia-master/builds/34661#018e3802-f274-4321-9018-f01d228c041d/916-1696

The question is - is this an alright change, or considered breaking? After all, it was undocumented that this threw at all...

vtjnash commented on base/task.jl:
```diff
@@ -356,8 +356,18 @@ function _wait2(t::Task, waiter::Task)
+"""
 function wait(t::Task; throw=true)
-    t === current_task() && error("deadlock detected: cannot wait on current task")
+    t === current_task() && throw(ArgumentError("deadlock detected: cannot wait on current task"))
```
ConcurrencyViolationError seems reasonable here. It is just what its name says: any detected errors with concurrency. It does not seem to be an ArgumentError.

Seelengrab committed:
Throw `ConcurrencyViolationError` in `wait(::Task)`

This always threw *something*, but now that it's documented to
throw, it's good to document *what* is thrown.

Seelengrab commented:
@vtjnash I'd like to add a docstring to `ConcurrencyViolationError` in a different PR, but noticed that pretty much nothing in `boot.jl` (where this is defined) has a docstring. Are docstrings for things defined in there added somewhere else due to bootstrapping issues with inline docstrings, or should I add it in there?

vtjnash commented:
base/docs/basedocs.jl

Seelengrab committed:
Ensure throwing the ConcurrencyViolationError works

The refactor needs `Core.throw` because `throw` is bound to the
local variable from the keyword.

Seelengrab commented:
Missed that there's a conflict between `Core.throw` and the `throw` keyword, should be good now. The question is though if the keyword should be renamed because of this conflict :thinking:

Seelengrab commented:
Is this good to go or should the keyword be renamed after all?
